File: src/agents/conversation-manager.ts
/**
 * Conversation Manager
 * 
 * Manages the conversation loop between agents and handles human-in-the-loop
 */

import { ComedyAgent } from "./comedy-agent.js";
import * as readline from "readline";

export interface HumanFeedback {
	type: "rating" | "comment" | "direction" | "stop";
	content: string;
	targetAgent?: string;
}

export class ConversationManager {
	private agents: ComedyAgent[];
	private iteration: number = 0;
	private maxIterations: number = 10;
	private humanFeedbackEnabled: boolean = false;
	private rl?: readline.Interface;

	constructor(agents: ComedyAgent[], maxIterations: number = 10) {
		this.agents = agents;
		this.maxIterations = maxIterations;
	}

	/**
	 * Start the conversation loop
	 */
	async start(): Promise<void> {
		console.log("\n🎭 Comedy Protocol - Multi-Agent Conversation");
		console.log("=" .repeat(50));
		console.log(`Starting conversation with ${this.agents.length} agents`);
		console.log(`Max iterations: ${this.maxIterations}`);
		console.log("=" .repeat(50) + "\n");

		// Run initial iterations without human feedback
		for (let i = 0; i < 3 && this.iteration < this.maxIterations; i++) {
			await this.runIteration();
		}

		// Enable human feedback after initial iterations
		this.humanFeedbackEnabled = true;
		console.log("\n👤 Human feedback enabled! You can now interact with the agents.\n");

		// Set up readline for human input
		this.rl = readline.createInterface({
			input: process.stdin,
			output: process.stdout,
		});

		// Continue with human-in-the-loop
		while (this.iteration < this.maxIterations) {
			await this.runIteration();
			
			// Ask for human feedback
			await this.requestHumanFeedback();
		}

		if (this.rl) {
			this.rl.close();
		}

		console.log("\n✅ Conversation complete!");
		this.printSummary();
	}

	/**
	 * Run one iteration of the conversation
	 */
	private async runIteration(): Promise<void> {
		this.iteration++;
		console.log(`\n${"=".repeat(50)}`);
		console.log(`ITERATION ${this.iteration}`);
		console.log(`${"=".repeat(50)}`);

		// Each agent takes a turn
		for (const agent of this.agents) {
			await agent.takeTurn(this.iteration);
			// Small delay between agents
			await new Promise((resolve) => setTimeout(resolve, 1000));
		}
	}

	/**
	 * Request human feedback
	 */
	private async requestHumanFeedback(): Promise<void> {
		if (!this.rl) {
			return;
		}

		return new Promise((resolve) => {
			try {
				console.log("\n👤 Human Feedback Options:");
				console.log("  [Enter] - Continue to next iteration");
				console.log("  'rate <agent> <1-10>' - Rate an agent's performance");
				console.log("  'comment <agent> <text>' - Comment on an agent");
				console.log("  'direction <text>' - Give direction to all agents");
				console.log("  'stop' - End conversation");
				console.log("  'summary' - Show conversation summary");

				this.rl!.question("\nYour input: ", async (answer) => {
					const feedback = this.parseHumanInput(answer);
					if (feedback) {
						await this.applyHumanFeedback(feedback);
					}
					resolve();
				});
			} catch (error) {
				// Readline might be closed, just continue
				resolve();
			}
		});
	}

	/**
	 * Parse human input into feedback
	 */
	private parseHumanInput(input: string): HumanFeedback | null {
		const trimmed = input.trim().toLowerCase();
		
		if (!trimmed || trimmed === "") {
			return null; // Continue
		}

		if (trimmed === "stop") {
			return { type: "stop", content: "Stop conversation" };
		}

		if (trimmed === "summary") {
			this.printSummary();
			return null;
		}

		const rateMatch = trimmed.match(/^rate\s+(\w+)\s+(\d+)$/);
		if (rateMatch) {
			return {
				type: "rating",
				content: rateMatch[2],
				targetAgent: rateMatch[1],
			};
		}

		const commentMatch = trimmed.match(/^comment\s+(\w+)\s+(.+)$/);
		if (commentMatch) {
			return {
				type: "comment",
				content: commentMatch[2],
				targetAgent: commentMatch[1],
			};
		}

		const directionMatch = trimmed.match(/^direction\s+(.+)$/);
		if (directionMatch) {
			return {
				type: "direction",
				content: directionMatch[1],
			};
		}

		return null;
	}

	/**
	 * Apply human feedback
	 */
	private async applyHumanFeedback(feedback: HumanFeedback): Promise<void> {
		if (feedback.type === "stop") {
			console.log("\n🛑 Stopping conversation per human request...");
			this.maxIterations = this.iteration; // Stop after current iteration
			return;
		}

		if (feedback.type === "direction") {
			console.log(`\n📢 Broadcasting direction to all agents: "${feedback.content}"`);
			// In a full implementation, you'd send this to agents
			// For now, just log it
			return;
		}

		if (feedback.targetAgent) {
			const agent = this.agents.find((a) => 
				a['config'].name.toLowerCase().includes(feedback.targetAgent!.toLowerCase())
			);
			
			if (agent) {
				console.log(`\n💬 Feedback for ${agent['config'].name}: ${feedback.content}`);
				// In a full implementation, you'd apply this feedback
			} else {
				console.log(`\n⚠️  Agent "${feedback.targetAgent}" not found`);
			}
		}
	}

	/**
	 * Print conversation summary
	 */
	private printSummary(): void {
		console.log("\n📊 Conversation Summary");
		console.log("=" .repeat(50));
		console.log(`Total iterations: ${this.iteration}`);
		console.log(`Agents: ${this.agents.length}`);
		
		for (const agent of this.agents) {
			const history = agent.getConversationHistory();
			console.log(`\n${agent['config'].name}:`);
			console.log(`  - Jokes contributed: ${history.length}`);
			console.log(`  - LLM: ${agent['config'].llmProvider}`);
		}
		
		console.log("\n" + "=".repeat(50));
	}
}

File: pyagents/conversation_manager.py
"""Conversation manager for the Python demo, mirroring the TypeScript flow."""

import asyncio
from typing import Iterable, List, Optional

from .agents import ComedyAgent


class ConversationManager:
	def __init__(self, agents: Iterable[ComedyAgent], max_iterations: int = 10):
		self.agents: List[ComedyAgent] = list(agents)
		self.max_iterations = max_iterations
		self.iteration = 0
		self.human_feedback_enabled = False

	async def start(self) -> None:
		print("\nComedy Protocol - Multi-Agent Conversation (Python)")
		print("=" * 50)
		print(f"Starting conversation with {len(self.agents)} agents")
		print(f"Max iterations: {self.max_iterations}")
		print("=" * 50 + "\n")

		# Run initial iterations without human feedback
		for _ in range(min(3, self.max_iterations)):
			await self.run_iteration()

		# Enable human feedback after initial iterations
		self.human_feedback_enabled = True
		print("\nHuman feedback enabled! You can now interact with the agents.\n")

		while self.iteration < self.max_iterations:
			await self.run_iteration()
			await self.request_human_feedback()

		print("\nConversation complete!")
		self.print_summary()

	async def run_iteration(self) -> None:
		self.iteration += 1
		print(f"\n{'=' * 50}")
		print(f"ITERATION {self.iteration}")
		print(f"{'=' * 50}")

		for agent in self.agents:
			await agent.take_turn(self.iteration)
			await asyncio.sleep(1)

	async def request_human_feedback(self) -> None:
		if not self.human_feedback_enabled:
			return

		print("\nHuman Feedback Options:")
		print("  [Enter] - Continue to next iteration")
		print("  'rate <agent> <1-10>' - Rate an agent's performance")
		print("  'comment <agent> <text>' - Comment on an agent")
		print("  'direction <text>' - Give direction to all agents")
		print("  'stop' - End conversation")
		print("  'summary' - Show conversation summary")

		# Use thread executor to avoid blocking the event loop
		loop = asyncio.get_event_loop()
		answer: str = await loop.run_in_executor(None, input, "\nYour input: ")
		feedback = self.parse_human_input(answer)
		if feedback:
			await self.apply_human_feedback(feedback)

	def parse_human_input(self, raw: str) -> Optional[dict]:
		trimmed = raw.strip()
		if not trimmed:
			return None

		lowered = trimmed.lower()
		if lowered == "stop":
			return {"type": "stop", "content": "Stop conversation"}

		if lowered == "summary":
			self.print_summary()
			return None

		if lowered.startswith("rate "):
			parts = lowered.split()
			if len(parts) == 3 and parts[2].isdigit():
				return {"type": "rating", "target": parts[1], "content": parts[2]}

		if lowered.startswith("comment "):
			parts = trimmed.split(maxsplit=2)
			if len(parts) == 3:
				return {"type": "comment", "target": parts[1], "content": parts[2]}

		if lowered.startswith("direction "):
			return {"type": "direction", "content": trimmed[len("direction "):].strip()}

		return None

	async def apply_human_feedback(self, feedback: dict) -> None:
		ftype = feedback.get("type")
		if ftype == "stop":
			print("\nStopping conversation per human request...")
			self.max_iterations = self.iteration  # stop after current iteration
			return

		if ftype == "direction":
			print(f"\nBroadcasting direction to all agents: \"{feedback.get('content')}\"")
			return

		target = feedback.get("target")
		if target:
			agent = next((a for a in self.agents if target.lower() in a.config.name.lower()), None)
			if agent:
				print(f"\n💬 Feedback for {agent.config.name}: {feedback.get('content')}")
			else:
				print(f"\n⚠️  Agent '{target}' not found")

	def print_summary(self) -> None:
		print("\n📊 Conversation Summary")
		print("=" * 50)
		print(f"Total iterations: {self.iteration}")
		print(f"Agents: {len(self.agents)}")

		for agent in self.agents:
			history = agent.get_conversation_history()
			print(f"\n{agent.config.name}:")
			print(f"  - Jokes contributed: {len(history)}")
			print(f"  - LLM: {agent.config.llm_provider}")

		print("\n" + "=" * 50)
